### ISSUE TITLE
Allow overriding naming convention

### ISSUE THREAD

**Issue by m-radzikowski** [serverless/serverless#9604]
### Use case description

Currently, the Framework generates the base name as `<service>-<stage>`. This base name is used for the CloudFormation stack name and as a prefix for various generated resources (functions, API Gateway, ...).

This naming may not be sufficient for all use cases, especially in bigger projects consisting of multiple stacks.

Currently, you can rename the individual resources like that:

```yml
provider:
  stackName: custom-stack-name # Use a custom name for the CloudFormation stack
  apiName: custom-api-name # Use a custom name for the API Gateway API
  websocketsApiName: custom-websockets-api-name # Use a custom name for the websockets API

functions:
  usersCreate:
    name: ${sls:stage}-lambdaName # optional, Deployed Lambda name
```

Still, internally SF assumes the default stack name, `<service>-<stage>`, is globally unique. This may not be true in all environments.

### Proposed solution

Introduce a parameter to override the naming strategy. It can be a top-level `commonPrefix` parameter. It should default to `<service>-<stage>`, but should allow to set custom naming, like `projectA-<stage>-<service>`.

### Related

- discussion here: https://github.com/serverless/serverless/issues/9361#issuecomment-860611172
- #2918
- #6598 (?)

**Comment by medikoo** [comment]
Thank you @m-radzikowski for proposal. It's definitely a worthwhile improvement.

Just to clarify, SF assumes that current naming prefix (which is internally configured as `<service>-<stage>` in most cases), is specific to given service + stage deployments. And that feels as right assumption.
Users should not configure two different services with same name - that rule should stay on a table, even after we address this proposal (e.g. in some places, as e.g. our dashboard we will still use just `service` to identify the service, and that won't change)

Use case we're addressing here, is that `<service>-<stage>` resource naming prefix, may not be best prefix for some cases. e.g. as explained in this [comment](https://github.com/serverless/serverless/issues/9361#issuecomment-860603931) in multi-service project, it might be more desirable to have setup as below:

```
service: <project>-<service>
commonPrefix: <project>-<stage>-<service>
```

Goal of this issue is to make that possible.

I'm just still not sure what would be the best name of a property through which we can override that. Candidates:
- `commonPrefix`
- `namePrefix`
- `name`
- `baseName` (my pick for now)

@pgrzesik @mnapoli what are your thoughts?

**Comment by mnapoli** [comment]
Hi @m-radzikowski, I'm curious: why would `projectA-<service>-<stage>` not work in your case?

The advantage with `projectA-<service>-<stage>` is that it is currently possible, and I would bet it is widely used today.

This is a pattern I use in my projects for example: I would have 2 different services named `port7777-api` and `port7777-website`. By convention, both services are forming the `port7777` application.

@medikoo isn't that as well how applications + services work in the Serverless Dashboard?

**Comment by medikoo** [comment]
> @medikoo isn't that as well how applications + services work in the Serverless Dashboard?

Yes, exactly. Check our exchange [here](https://github.com/serverless/serverless/issues/9361#issuecomment-860595386), I raised same questions, and @m-radzikowski explained use case in more detail.

**Comment by mnapoli** [comment]
I see, thank you. From what I gather, the motivation behind the change is:

> To be able to filter resources by project + stage (_not the service_) in the AWS Console.

Let me know if I'm missing another use case.

I was wondering if it made sense to allow setting an `app` name in `serverless.yml` without requiring the Serverless Dashboard. But even then, the Dashboard works `app -> service -> stage`, not the other way around, so that wouldn't solve that specific issue.

**Comment by m-radzikowski** [comment]
> To be able to filter resources by project + stage (not the service) in the AWS Console.

Yes, this is my motivation. More generally, to set naming schema in a single place, instead of overriding the stack, function, and API Gtw. names individually. This should enable anyone to use the kind of naming strategy they want.

I understand that the Serverless Dashboard requires the service name to be unique. From this point of view, in monorepo (multi-stack) approach, it makes the most sense to just prepend the service name with some "project name", like so:

```
service: projectA-serviceX
```

And this would work well with Serverless Dashboard.

However, the CF stack naming `projectA-serviceX-stageQ` and resource naming `projectA-serviceX-stageQ-resource1` is not really logical for me (it may be for you, I won't argue on that!). Especially with AWS Console where filtering works by the name prefix in many places.

I usually want to look at the resources for `projectA` and stage `stageQ`, not `projectA` and `serviceX` (that would include many different stages). Especially, I want to look at all resources on `prod` (let's say the CloudWatch Alarms), not mixed up with resources from other stages.

Again, I understand that the Serverless Dashboard works the other way around. But:

1. Changing the resources naming is possible right now, it just can't be done in a single place, but for each resource individually.
2. Serverless Dashboard requires the `service` name to be unique. But the Serverless Framework itself does not, I think? Apart from the problem with #9361. Basically, if the resource names do not clash, there should be no problem with a deployment. Would you be ok to leave that to the users, so if we don't use Serverless Dashboard and we reuse service names but override stack names we are still ok? This is what works well for us at the moment.

As for the proposed parameter name - I like the `baseName`. In fact, this is what we do right now:

```yml
custom:
  project: ${file(../config.yml):project}
  baseName: ${self:custom.project}-${opt:stage}-${self:service}

provider:
  name: aws
  stackName: ${self:custom.baseName}
```

**Comment by medikoo** [comment]
> Serverless Dashboard requires the service name to be unique. But the Serverless Framework itself does not

It does. Deploying two different services with same name will simply fail  (it works in your case only because you hack some internal names, but as you can see it's imperfect, as you cannot cover all naming cases that way)

Maybe issue is that it's not explicitly stated in a documentation, it definitely should.

> As for the proposed parameter name - I like the baseName.

Great. I propose to stick to that then

**Comment by m-radzikowski** [comment]
> > Serverless Dashboard requires the service name to be unique. But the Serverless Framework itself does not
> 
> It does. Deploying two different services with same name will simply fail (it works in your case only because you hack some internal names, but as you can see it's imperfect, as you cannot cover all naming cases that way)

So won't changing the `baseName` enable it then? With the user being responsible for the uniqueness of the `baseName` then, instead of `service` name.

My understanding:

- now the `service` name must be unique because the implicit `baseName` is `<service>-<stage>`
- after the change, the `baseName` would need to be unique
  - that means if you don't override the `baseName`, still the `service` name must be unique

**Comment by medikoo** [comment]
> So won't changing the baseName enable it then? 

No, it wouldn't. e.g, in Serverless Dashboard you'd still have the collision, and we should not assume it won't affect some other places.

This feature is not about enabling to use same `service` names across different services (imo rule of `service` name to be unique should by all means remain on a table)

What we're addressing here is to allow customization of general naming pattern for generated resources, that's it.

**Comment by pgrzesik** [comment]
While I think it's definitely a worthwhile change to introduce `baseName` for some use cases, I'm wondering if, at least for AWS, the ability to filter out specific resources quickly in AWS console isn't better achieved with the use of `tags` or `stackTags`. Is there a reason why you prefer to use `name`-based filtering instead of grouping services logically into a single project via tags?

**Comment by m-radzikowski** [comment]
> While I think it's definitely a worthwhile change to introduce `baseName` for some use cases, I'm wondering if, at least for AWS, the ability to filter out specific resources quickly in AWS console isn't better achieved with the use of `tags` or `stackTags`. Is there a reason why you prefer to use `name`-based filtering instead of grouping services logically into a single project via tags?

If only AWS would be consistent across the whole Console, and all resources would inherit tags from CF. I know there are Resource Groups, but it's somehow easier for me to just go to the service where I want to find a particular resource.

**Comment by medikoo** [comment]
> what about `resourcePrefix` as a config key?

This prefix affects a bit more than resource names, e.g. it will also influence CF stack name or output names.

At this point `baseName` feels as best pick to me.

**Comment by dmeiser** [comment]
We're currently unable to upgrade because it would require us to rename too many resources - at some point during the 2.30 series, outputs were added. I was coming to file a bug against the default naming behavior because it is too difficult to override all resource and output names.

We create multiple environments in the same account. Sometimes those environments have the same stage names (eg - 2 prod stacks with different base names). We would preferably like to append our resources with with <resource-name>-<arbitrary-name>-<stage>.

This would also let us optionally add a region identifier, if we have need to go multi-region.

**Comment by medikoo** [comment]
I've updated main issue description with implementation spec. PR's welcome!

**Comment by ktwbc** [comment]
Was there ever any movement on this? Our org uses the environment name for organization and IAM permissions, so putting prod and dev as the stack name before the service instead of after means you can grant access to an arn ending in dev-* or prod-* simply because permissions handle prefixes but not suffixes.

**Comment by dgarson** [comment]
We are trying to deploy the same Lambda (`service`) to multiple VPCs within the same AWS account/region. Unfortunately, Serverless does not support overriding the top-level `service` property (or including any variable substitution, which would also work). This change would also suffice to satisfy this use case, and may in fact be a much more preferable option in a majority of these cases. This would likely introduce redundant `-{stage}` in naming, but any support at all for this use case would be fantastic at this point.

I've begun taking a stab at this and will probably require some additional information/context surrounding the ramifications of using such a `baseName` for virtually all occurrences of the `service`. Since I am newer to the codebase, I thought this may be an acceptable solution since anyone overriding the `baseName` will be knowingly diverging from their current configuration, and I am concerned that if we do not reuse this everywhere, that we will end up encountering other negative side effects due to the presumption that the `service` name is unique.

I'm currently using an approach where 2 variables are introduced, rather than just the `baseName`. I've also added a `shortName` which is either the provided `baseName` (as-is), or it is the `service` name (without any `-{stage}`). This is useful for the cases where the resource naming is inverted, e.g. some API gateway always use `{stage}-{service}`, to avoid redundant `-{stage}` when not overriding `baseName`. Similarly, the `shortName` can be used for things like Dashboard names/descriptions to preserve complete backwards-compatibility in config/output if users were to omit the new `baseName`. If we simply use the `service` name for things like this, then the dashboards will have a name/description that is overly-broad for the deployment of that `service`.

Any thoughts/concerns would be greatly appreciated!
